app/src/features/apiClient/screens/apiClient/utils.ts

```diff
@@ -184,3 +199,113 @@ export const convertFlatRecordsToNestedRecords = (records: RQAPI.Record[]) => {
+    [AUTHORIZATION_TYPES.API_KEY]: { key: "", value: "", addTo: "HEADER" },
+    [AUTHORIZATION_TYPES.BEARER_TOKEN]: { bearer: "" },
+    [AUTHORIZATION_TYPES.BASIC_AUTH]: { username: "", password: "" },
+  };
+};
+
+export const updateAuthOptions = (
+  entry: RQAPI.Entry,
+  currentAuthType: AUTHORIZATION_TYPES,
+  updatedKey?: string,
+  updatedValue?: string
+) => {
+  // clear all based details on the current entry:
+  const currentEntry = { ...entry };
+  const oldAuth = currentEntry.auth ?? {
+    currentAuthType: AUTHORIZATION_TYPES.NO_AUTH,
+    authOptions: getEmptyAuthOptions(),
+  };
+  const newAuthOptions =
+    updatedKey || updatedValue
+      ? {
+          ...oldAuth.authOptions,
+          [currentAuthType]: {
+            ...oldAuth.authOptions[currentAuthType],
+            [updatedKey]: updatedValue,
+            ...(updatedKey || updatedValue ? { [updatedKey]: updatedValue } : {}),
+          },
+        }
+      : oldAuth.authOptions;
+
+  currentEntry.auth = {
+    currentAuthType,
+    authOptions: newAuthOptions,
+  };
+
+  !isEmpty(currentEntry.request.headers) && deleteAuthHeaders(currentEntry);
+  !isEmpty(currentEntry.request.queryParams) && deleteAuthQueryParams(currentEntry);
+
+  let newKeyValuePair: KeyValuePair;
+
+  const createAuthorizationHeader = (type: string, key: string, value: string): typeof newKeyValuePair => ({
+    id: Math.random(),
+    key,
+    value,
+    type,
+    isEnabled: true,
+  });
+
+  const updateDataInState = (data: KeyValuePair[], type = AUTH_ENTRY_IDENTIFIER, key: string, value: string) => {
+    const existingIndex = data.findIndex((header) => header.type === type);
+    const newHeader = createAuthorizationHeader(type, key, value);
+
+    if (existingIndex !== -1) {
+      data[existingIndex] = { ...data[existingIndex], ...newHeader };
+    } else {
+      data.unshift(newHeader);
+    }
+  };
+
+  switch (currentAuthType) {
+    case AUTHORIZATION_TYPES.NO_AUTH:
+      break;
+    case AUTHORIZATION_TYPES.BASIC_AUTH: {
+      const { username, password } = newAuthOptions[currentAuthType];
+      updateDataInState(
+        currentEntry.request.headers,
+        AUTH_ENTRY_IDENTIFIER,
+        "Authorization",
+        `Basic ${btoa(`${username || ""}:${password || ""}`)}`
+      );
+      break;
+    }
+    case AUTHORIZATION_TYPES.BEARER_TOKEN: {
+      const { bearer } = newAuthOptions[currentAuthType];
+      updateDataInState(currentEntry.request.headers, AUTH_ENTRY_IDENTIFIER, "Authorization", `Bearer ${bearer}`);
+      break;
+    }
+    case AUTHORIZATION_TYPES.API_KEY: {
+      const { key, value, addTo } = newAuthOptions[currentAuthType];
+
+      if (key) {
+        updateDataInState(
+          addTo === "QUERY" ? currentEntry.request.queryParams : currentEntry.request.headers,
+          AUTH_ENTRY_IDENTIFIER,
+          key || "",
+          value || ""
+        );
+      }
+      break;
+    }
+    default:
+      break;
+  }
```

Comment:
The addition of auth header should be a runtime logic. They shouldn't be added into the entry during the saving at all (although they are getting removed at the end in sanitise entry).

Reply:
This generation of headers should be somewhere over [here](https://github.com/requestly/requestly/blob/0ac721b84b5c2a31ac3ea0655ef88a30aff3f326/app/src/features/apiClient/helpers/APIClientManager/index.ts#L17) so that it doesn't effect the final entry which is saved in the db.

For showing in the Headers Panel, we should consume the same function which generated the headers and then should it separately which changing anything in the entry